### ISSUE TITLE
Make the updated ACL take effect without disconnecting

### ISSUE THREAD

**Issue by Explorer0** [vernemq/vernemq#1935]
Is there any way to make the **updated (subscribe/publish)ACL** (storaged in mysql)take effect without disconnecting?

**Comment by ioolkos** [comment]
No, this is not implemented. There's an exported internal function `clear_cache/2` here, which might be a starting point to look into this: https://github.com/vernemq/vernemq/blob/master/apps/vmq_diversity/src/vmq_diversity_cache.erl

In any case, it will always have to be externally triggered.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by Explorer0** [comment]
> No, this is not implemented. There's an exported internal function `clear_cache/2` here, which might be a starting point to look into this: https://github.com/vernemq/vernemq/blob/master/apps/vmq_diversity/src/vmq_diversity_cache.erl
> 
> In any case, it will always have to be externally triggered.
> 
> 👉 Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq 👉 Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

Thanks, but how to use `clear_cache/2` to clean session ACL cache in vernemq server ?

**Comment by mths1** [comment]
I start to feel to need this as well. My suggestion would be to have a new hook (like invalidate_cache) that can be externaly triggered e.g. via vmq-admin or the api. In combination with the already existing reauth session all the needed functionality would be provided.

**Comment by ioolkos** [comment]
The caches are implemented in the respective plugins. The cache implementation will provide the invalidate functions. This can be triggered externally under the `vmq-admin` command for the specific plugin. Hooks grab internal events and allow you to implement what should happen; so I guess here, a new hook is not even needed.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.

**Comment by mths1** [comment]
The idea with the hooks would have been that vmq-admin does not need to know any specifics of the plugin implementation (incl. third-party, community provided).

**Comment by ioolkos** [comment]
@mths1 Ok, I see what you mean. If we can find a nice direct approach, we might as well go for the generic functionality here.
Let's give it some thought & discussion.

---
:point_right: Thank you for supporting VerneMQ: https://github.com/sponsors/vernemq
:point_right: Using the binary VerneMQ packages commercially (.deb/.rpm/Docker) requires a paid subscription.
